google/cloud/internal/future_void.h

```diff
@@ -63,10 +64,90 @@ class future<void> final : private internal::future_base<void> {
   using future_base::wait_for;
   using future_base::wait_until;
 
+  /**
+   * Attach a continuation to the future.
+   *
+   * Attach a callable @a func to be invoked when the future is
+   * ready.  The return type is a future wrapping the return type of
+   * @a func.
+   *
+   * @return future<T> where T is std::result_of_t<F, R> (basically).
+   * If T matches future<U> then it returns future<U>.  The returned
+   * future will contain the result of @a func.
+   * @param func a Callable to be invoked when the future is ready.
+   * The function might be called immediately, e.g., if the future is
+   * ready.
+   *
+   * Side effects: valid() == false if the operation is successful.
+   */
+  template <typename F>
+  typename internal::then_helper<F, void>::future_t then(F&& func) {
+    check_valid();
+    using requires_unwrap_t =
+        typename internal::then_helper<F, void>::requires_unwrap_t;
+    return then_impl(std::forward<F>(func), requires_unwrap_t{});
+  }
+
+  /**
+   * Return true if the future is satisfied.
+   *
+   * @throws std::future_error if the future is invalid.
+   */
+  bool is_ready() const {
+    check_valid();
+    return shared_state_->is_ready();
+  }
+
  private:
   friend class promise<void>;
   explicit future(std::shared_ptr<shared_state_type> state)
       : future_base<void>(std::move(state)) {}
+
+  /// Implement `then()` if the result does not require unwrapping.
+  template <typename F>
+  typename internal::then_helper<F, void>::future_t then_impl(F&& functor,
+                                                              std::false_type) {
+    // g++-4.9 gets confused about the use of a protected type alias here, so
```

Comment:
What is the minimum supported g++ version?

Reply:
That is poorly defined right now. The CI builds have a gcc-4.9 build enabled.  We decided to support the distributions that GCE supports, which right now include Ubuntu 14.04 and RHEL 7.  Their default compilers are gcc-4.8 based, which we could not get to work, so we sort of settled for 4.9.  We may need to  reconsider that decision (#798).